MoreLinq/Memoize.cs

```diff
@@ -0,0 +1,136 @@
+        /// in-memory source.
+        /// An additional argument specifies if buffering must happen even if the source implements
+        /// <see cref="ICollection{T}"/> (otherwise it is assumed that the source already serves as 
+        /// a buffer adequately).
+        /// </summary>
+        /// <param name="source">The source sequence.</param>
+        /// <param name="forceBuffering">Force buffering, even if source is an <see cref="ICollection{T}" />
+        /// (otherwise source is assumed to already be in-memory and enumerated quickly, and hence that
+        /// buffering again would be a waste).</param>
+        /// <returns>A <see cref="MemoizedEnumerable{T}"/>.</returns>
+        public static IBufferedEnumerable<T> Memoize<T>(this IEnumerable<T> source, bool forceBuffering)
+        {
+            if (source == null) throw new ArgumentNullException("source");
+
+            return (source as MemoizedEnumerable<T>) ?? new MemoizedEnumerable<T>(source, forceBuffering);
+        }
+    }
+
+    /// <summary>
+    /// An enumeration that, when being iterated, lazily iterates through the source enumeration 
+    /// and cache the elements already iterated for future iterations. It supports partial iterations.
+    /// </summary>
+    /// <typeparam name="T">Type of the source sequence</typeparam>
+    public interface IBufferedEnumerable<T> : IEnumerable<T>
+    {
+
+    }
+
+    internal class MemoizedEnumerable<T> : IBufferedEnumerable<T>, IEnumerable<T>
+    {
+        private ICollection<T> collection;
+        private IEnumerable<T> source;
+        private IEnumerator<T> e;
+        private IList<T> cache;
+        private bool disposed;
+
+        public MemoizedEnumerable(IEnumerable<T> sequence, bool forceBuffering)
+        {
+            if (sequence == null) throw new ArgumentNullException("sequence");
+		
+            if (!forceBuffering && sequence is ICollection<T>)
+            {
+                collection = (ICollection<T>)sequence;
+            }
+            else
+            {
+                source = sequence;
+                cache = new List<T>();
+            }
+        }
+
+        private IEnumerator<T> GetMemoizedEnumerator()
+        {
+            if (e == null && !disposed) e = source.GetEnumerator();
+
+            int index = 0;
+
+            while (true)
+            {
+                while (index < cache.Count)
+                {
+                    yield return cache[index++];
+                }
+
+                if (!disposed && e.MoveNext())
+                {
+                    cache.Add(e.Current);
+                    index++;
+                    yield return e.Current;
+                }
+                else
+                {
+                    if (!disposed)
+                    {
+                        disposed = true;
+                        e.Dispose();
+                        source = null;
+                        e = null;
+                    }
```

Comment:
> How someone would be able to use `using (var m = TestSequence().Memoize())` if MemoizeEnumerable is not exposed and Memoize returns IEnumerable<T>, which doesn't implements `IDisposable` itself?

You can't use it without casting to `IDisposable` explicitly. You cannot return `MemoizeEnumerable` from `Memoize` because when `Memoize` is given a collection, it returns it directly and so `IEnumerable<T>` remains the common type between the two.

> So user is forced to added a whole packaged as dependence only because he or she wants to use a single method of it? And all this method do is a simple/common chain operator? As user, I would not be happy with this scenario.

What's wrong with that? The dependency is a distribution cost. It's not a runtime cost. At runtime, you pay JIT tax for only the code path you use. Those who want to remove the distribution cost can either use IL-merging or do source-inclusion. MoreLINQ has the same issue. In some projects, I may need only one or two operator. Should I then not use MoreLINQ? Everyone's mileage may vary.

> > As the implementation currently stands, the cache will live as long as you hold on to the enumerable

> Is it wrong? If yes, why?

What's wrong is that the cache will live even _after the disposal_.

> For use the `using` statement the type must implements IDisposable, not simply have a Dispose method.

Correct and so you must cast the result of `Memoize` explicitly to find out if it supports disposing.

> How about if `source as IDisposable` is `null` ?

It's harmless. C#'s `using` statement generates code that makes a null check in the `finally` block.